### PR TITLE
Add pre-commit hook for prettier and eslint --fix

### PR DESCRIPTION
**Pull request by lukebelliveau**
**tl;dr:** All commits must pass `yarn prettier` and `yarn eslint --fix` in the `src/` folder.

This PR adds a `format` script to run `prettier` and `eslint --fix` on the `src/` folder, and uses `lint-staged` and `husky` to run it before each commit.  It also switches the `eslint` config to use `eslint-config-prettier` rather than `eslint-plugin-prettier`, now that `prettier` is run on each commit with this merge and no longer needs a check by `eslint`. Discussed in [#1372](https://github.com/styled-components/styled-components/issues/1372).

I intentionally circumvented the pre-commit hook here because I anticipate feedback on the configuration or targets. Before this is merged, we should push another commit to run `yarn format` on the code.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -1,7 +1,10 @@
 {
   "parser": "babel-eslint",
-  "extends": "airbnb",
-  "plugins": ["flowtype", "flowtype-errors", "prettier"],
+  "extends": [
+    "airbnb",
+    "prettier"
```
Comment (philpl) [inline .eslintrc]: Several things here; this is a nice change, but we should use the following config too to ensure smooth compatibility with flow: `"prettier/flowtype"`

Please do not remove the prettier plugin (see below)

File: package.json [hunk 2]
```diff
@@ -24,9 +25,11 @@
     "test:size": "bundlesize",
     "flow": "flow check",
     "flow:watch": "flow-watch",
+    "format": "yarn run prettier && eslint src --fix",
```
Comment (philpl) [inline package.json]: It's a big advantage to run prettier as part of the eslint plugin for multiple reasons:

- Integration of the two
- Avoid duplicated/staggered errors
- Avoid confusing order of prettier & eslint (here eslint runs after prettier, so there's still a chance of conflicts if we change anything in the eslint config)
- Integration into editors that have eslint integration but not prettier integration
- And a big one: get all messages in the same formatted way

Comment (lukebelliveau) [reply package.json]: Thanks for the feedback! To ensure I understand, we want to:

- Re-add `eslint-plugin-prettier`, exactly as was implemented by @existentialism 
- Set the pre-commit hook to run `eslint src --fix`. With `eslint-plugin-prettier`, this script will automatically run `prettier`
- Add `eslint-plugin-flowtype` to .eslintrc

Is this correct?

Comment (philpl) [reply package.json]: > Add eslint-plugin-flowtype to .eslintrc

The plugins should be there, rather we'll want to add the eslint-config-prettier/flow config. There's more info on the available configs here: https://github.com/prettier/eslint-config-prettier

> If so - we still need to keep eslint-config-prettier, right?

That'd be great, yea :+1:

File: .eslintrc [hunk 3]
```diff
@@ -2,9 +2,10 @@
   "parser": "babel-eslint",
   "extends": [
     "airbnb",
-    "prettier"
+    "prettier",
+    "prettier/flowtype"
```
Comment (philpl) [inline .eslintrc]: 👌

**Comment by existentialism** [comment]
Can drop `pre-commit` dep and the pre-commit key from package.json since we're replacing it with husky?

**Comment by lukebelliveau** [comment]
Sorry, I'm new to this collaborator thing :). Looks like my push addressing @existentialism's suggestion automatically dismissed @philpl's review - if this was bad etiquette, I'd really appreciate feedback on how to do this right!

File: package.json [hunk 4]
```diff
@@ -145,11 +148,10 @@
   },
   "lint-staged": {
     "*.js": [
-      "eslint --fix",
+      "format",
```
Comment (glennreyes) [inline package.json]: Will this format/lint the whole codebase before each commit?
I think what we rather want here is to replace this with
```
"prettier --write",
"eslint --fix"
```
... so that only staged files are formatted.

File: .eslintrc [hunk 5]
```diff
@@ -1,6 +1,10 @@
 {
   "parser": "babel-eslint",
-  "extends": "airbnb",
+  "extends": [
+    "airbnb",
+    "prettier",
+    "prettier/flowtype"
+  ],
   "plugins": ["flowtype", "flowtype-errors", "prettier"],
```
Comment (glennreyes) [inline .eslintrc]: IIRC, we can ditch prettier plugin (which runs it as eslint rule) if we use prettier as standalone?

Comment (lukebelliveau) [reply .eslintrc]: That's correct, but it's preferred to run it as part of `eslint` per @philpl's comment [here](https://github.com/styled-components/styled-components/pull/1376#discussion_r157919895).

File: package.json (hunk 4 above)
Comment (lukebelliveau) [reply package.json]: Hmm.. I just implemented this change locally, and now neither solution is running the lint-staged scripts on commit. Here's what I'm doing:

1. Make a change in a file in the `src` folder that's nonconformant to our prettier rules (like adding an extra tab or a really long line).
2. `git add src && git commit -m "This should run the 'format' command"`
3. I see the files immediately committed, and nothing from `husky`.

I've clearly gone mad, so I'm gonna go to bed :P. Is anybody else seeing this behavior?

File: package.json [hunk 6]
```diff
@@ -145,11 +148,10 @@
   },
   "lint-staged": {
     "*.js": [
-      "eslint --fix",
```
Comment (philpl) [inline package.json]: Can you revert this to just `eslint --fix` please? Otherwise lint-staged is not going to be able to pass the files to eslint, since the format script is already carrying some filenames.

File: package.json [hunk 7]
```diff
@@ -24,9 +24,11 @@
     "test:size": "bundlesize",
     "flow": "flow check",
     "flow:watch": "flow-watch",
+    "format": "eslint src --fix",
```
Comment (philpl) [inline package.json]: Can you change `src` here to `./**/*.js` so that it targets the same files that lint-staged is targeting please?

**Comment by lukebelliveau** [comment]
@philpl I pushed the requested changes in case I'm just having an issue with my machine, but somebody should probably manually verify this actually works before approving - [the hook doesn't actually seem to be working now](https://github.com/styled-components/styled-components/pull/1376#discussion_r157974126).

This _was_ working for me last night, I'm feeling loopy over this! I'll keep trying to chase it down.

**Comment by philpl** [comment]
@lukebelliveau sorry for the (festive) delay 😅 can you run “eslint —fix” on the entire src folder please? Just to be sure

**Comment by lukebelliveau** [comment]
@philpl No worries, happy holidays :smiley::christmas_tree:!

The linter was run before my last commit. I just double checked and there were no changes for `eslint --fix` (or `prettier`, for that matter). Am I missing something?

**Comment by philpl** [comment]
@lukebelliveau no worries, that seems alright 👌 it would also not be bad if we missed a file, since the next person editing it would trigger the hook again on it

**Comment by lukebelliveau** [comment]
Oops @philpl, I totally forgot to add this to CHANGELOG.md. I [pushed a change](https://github.com/styled-components/styled-components/commit/1ac97e343d8e631a79405f33fc4f9d023c72bb65) just now, but want to make sure I'm going about this the right way before opening another PR. Shall I open another?

**Comment by philpl** [comment]
@lukebelliveau sure, let’s just open another PR; sorry for the trouble 😅

**Comment by lukebelliveau** [comment]
Cool, done [here](https://github.com/styled-components/styled-components/pull/1393/). Sorry about that!!
